### DISCUSSION TITLE
clashtui 的 profile 统一使用 yaml 格式?

### DISCUSSION THREAD

**Discussion by JohanChane** [JohanChane/clashtui#29] [Ideas]
clashtui 的 profile 统一使用 yaml 格式, 将 mihomo 的 profile 放到 `profile_cache`, 这样可以为 clashtui profile 添加一些选项:
- 是否使代理更新
- 添加转换工具: 比如过滤节点; 批量更改名称; 将 `proxy-providers` 放到 `proxies` 字段; 
    - 转换过程由用户自己编程。clashtui 只提供一些常用的接口。

**Comment by Jackhr-arch** [comment]
## About Profiles
Now I'm trying to classify them as `File`(from local file), `Url` and `Generated`(from template), and store them in config file, while those yaml files in `path/profile`. I believe this can make management easier.
Current I set `Generated` as upgradable(using same function as `Url`), maybe we can do some more thing for this type.
`File` is not upgradable as we do not store the path

**Reply by JohanChane** [reply]
我之前也有这样的想法

**Comment by JohanChane** [comment]
为了表达更加清晰, 我继续用中文讨论。

我之前一直在想这个功能如何设计, 但是我想实现的东西太多导致设计很混乱。

我是想设计一种 clashtui 专门使用的 profile, 比如:

```
clashtui: true      # 用于区分 yaml profile (或许可以不用)

proxy-providers:
-   <name>:
        <url>: 
        {type: github/gitlab/..., <token>: }
        <filter>: 转换程序
        <proxy>: 是否使用代理更新

proxy-groups:
-   select:
-   auto:
-   ...

pre_rules:
```

解释上面的配置:
-   token: See [ref](https://stackoverflow.com/a/22314262/10498412)
-   filter:  从 stdin 读入配置, 将转换后的结果输出到 stdout。clashtui 将其 stdout 写入到相应的文件。
-   proxy: 有些机场需要直连更新。
-   有些人喜欢用机场的配置:
    -   pre_rules: 添加自己的规则。
    -   proxy-groups: 添加 select 组, auto 组, ...

灵活性一般会牺牲方便性, 使用这个配置会更加灵活, 但是配置麻烦。所以之前的 profile 类型也会保留 (可以修改设计)。

**Reply by Jackhr-arch** [reply]
you mean template? As I think, the profile generated from template should be able to do a 'upgrade'(regenerate from original template, including re-sync related profiles and turn proxy provider into proxy or something).

**Reply by JohanChane** [reply]
是的, 我想不要 filter, pre_rules, proxy-groups 了, 感觉与模板的功能重叠了。基本上有两个主要功能:
-   url (proxy-provider 和 profile url): 可以指定 `{url:, type:, token: }`, clashtui 能下载 url 对应的文件即可。
-   添加模板的新功能 (解决用户想保留机场的节点分组与分流):
    -   是否使用 `proxy-provider` 的所有 `proxy-groups` (插入到 `proxy-groups` 的后面即可。) 。使用这个功能一般只有一个 proxy-provider, 不要考虑 proxies 的重名问题了。如果要解决重名问题是比较麻烦的 (可能强制 proxies 添加前缀避免重名)。
    -   是否使用 `proxy-provider` 的所有 rules (插入到 `rules` 的后面即可。这样自定义的规则会在前面)

**Reply by Jackhr-arch** [reply]
Just add what you want, there is no rule saying that we must follow Mihomo's profile schema.

**Comment by JohanChane** [comment]
这个可能是 clashtui 内部使用统一的数据结构 (clashtui profile), 对于用户而言, 之前的 profile 类型尽量可以使用。clashtui profile 的设计关联的新功能比较多, 我目前还没有想得很细, 我会过一段时间再做了, 你先挑你感兴趣的功能做。或者按照你的 clashtui.conf 来做, 如果改动的话, 也只是修改用户使用习惯的代码, 实现功能的代码不需要改。

**Comment by JohanChane** [comment]
## url 格式

有两种格式:

```yaml
url: "<url>"    # 旧的格式
url: {type: '<type>', url: '<blob file url>', token: '<token>'}     # 新的格式
```

url 格式的应用范围:
-   profile url
-   proxy-provider url

## profile 的类型

-   SubUrl: 只一个订阅链接。
-   ClashPf: clash 支持的 profile。是一个 yaml 文件。
-   CtPf: ClashTui 的 profile 格式。是一个 yaml 文件, 且包含 `clashtui` section。
-   CtPfWithSubUrl: 和 `CtPf` 区别不大。是一个 yaml 文件。且包含 `clashtui: {profile_url: ...}`

See [UrlItem](https://github.com/JohanChane/clashtui/blob/3fc73c0a0ecaa6752f9bf69fd1abda9cc74428dc/clashtui/api/src/clash.rs#L30), [ProfileItem](https://github.com/JohanChane/clashtui/blob/3fc73c0a0ecaa6752f9bf69fd1abda9cc74428dc/clashtui/src/utils/tui.rs#L25)

for example: CtPfWithSubUrl

```yaml
clashtui:
  profile_url: {type: 'github', url: '<blob file url>', token: '<token>'}     # 目前只支持 github
```

See [github token](https://stackoverflow.com/a/78687474/10498412)

有些 REST API 下载文件比较麻烦:
-   取得 file blob sha
-   拼接 blob path
-   取得响应包的 content, 并使用 `base64 -d` 取得文件内容。

See more:
-   [gitee](https://gitee.com/api/v5/swagger#/getV5ReposOwnerRepoGitBlobsSha)
-   [gitcode](https://docs.gitcode.com/docs/openapi/repos/)

### ClashTui profile 的不同点

proxy-provider 的选项 (还没有完成):
-   prepend_rules           # 保留 proxy_provider 的东西
-   appdend_rules
-   prepend_proxy_groups
-   appdend_proxy_groups

不使用代理更新 profile:
-   使用新的 url 格式的 `with_proxy` 即可。

如果有 ClashTui 有新的 url 格式, 则在合并 profile 到 `config.yaml` 时, 只有使用 `no proxy provider` 选项。

或者如果所有都能生成一个含有 token 的 url, 则不用使用 `no proxy provider` 选项。比如:

```
github url format:
- 'https://token@url_without_protocol'
- 'https://x-access-token:token@url_without_protocol'
e.g. 'https://<token>@raw.githubusercontent.com/<User>/<Repo>/<Branch>/config.yaml'
```

See [gen_url](https://github.com/JohanChane/clashtui/blob/344b0a1261921d0182d9fe94feb297b3b950c4d9/clashtui/api/src/clash.rs#L69)

似乎 minreq 并不支持含有 token 的 url, 但是 reqwest 是支持的。

---

目前想到只有这些, 可以随时修改。

**Comment by Jackhr-arch** [comment]
抱歉，现在才看到这部分

关于Github Token，根据[Github Restful API](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#create-a-repository-for-the-authenticated-user)，或你给出的[StackOverflow](https://stackoverflow.com/a/78687474/10498412)，只需要在`header`中添加 `"Authorization: Bearer <YOUR-TOKEN>"`即可

更好的消息是，这里早已有[一个实现](https://github.com/JohanChane/clashtui/blob/0e9a58dbe79200f41bbcc69af45b5cb00cd5b699/src/webapi.rs#L49)。它在这里是用来实现到[clash内核的验证](https://wiki.metacubex.one/api/#_1)。

只需稍作修改，例如在`mock_clash_core`中追加一个`with_token`，或者单独实现一个`Url`结构体，对它的适配很快就能完成

**Comment by Jackhr-arch** [comment]
### 关于clashtui专有profile的格式
- 使用`clashtui:`是一个很好的方式。
- 区分`CtPf`是不必要的，可以选择在下载一个profile后继续解析它，即递归
- 我更建议将 proxy_provider的保留与否 作为`clashtui:`下的一个配置项

### 建议
最终样式
```yaml
clashtui:
  with_proxy: null | true | false
  replace_proxy_provider: null | true | false
  url:
  - !Github
    url: !ClashPf ''
    token: ''
  - !Single
    url: ''
```
对应的结构
```rust
#[derive(serde::Deserialize)]
enum TuiUrlItem {
    Github { url: TuiUrl, token: String },
    Single { url: String },
}
#[derive(serde::Deserialize)]
enum TuiUrl {
    ClashPf(String),
    ClashTuiPf(String),
}
#[derive(serde::Deserialize)]
struct ClashTuipf {
    with_proxy: Option<bool>,
    replace_proxy_provider: Option<bool>,
    url: Vec<TuiUrlItem>,
}


let mut the_yaml = serde_yaml::Mapping::new();
if let Some(tuipf) = the_yaml
    .remove("clashtui")
    .and_then(|v| serde_yaml::from_value::<ClashTuipf>(v).ok())
{
    // do something
}

**Comment by Jackhr-arch** [comment]
### 关于实现模板过程中遇到的一些问题
在使用默认模板，即使用[这个模板](https://github.com/JohanChane/clashtui/blob/20b068689d2f31bf40a7c40d0da881b280e13c18/Doc/my/templates/my_tpl.yaml)验证代码时，使用版本为v0.2.3-main-20b0686-debug（即目前main分支）的程序会在src/utils/tui/impl_profile.rs:217:71处panic

环境为全默认，在clashtui初始化后仅在/template文件夹下创建template_proxy_providers与上述模板文件。

替换其他模板后，问题依旧

我希望你能明确模板的设计，为避免概念变得太复杂，我们暂且仅考虑从模板到配置文件这一过程 @JohanChane

**Reply by JohanChane** [reply]
哥们, 你是真的喜欢编程啊, 都快除夕了, 你不忙的吗。

template_proxy_providers 不能为空。这应该是一个 bug。

目前的模板很灵活, 但是确实复杂了。从用户的角度来说, 他们一般有这样的需要:
- 为某个 proxy-provider 添加一个 auto/select/fallback 组这样。
你也可以重新设计这块的功能, 尽量简单。之后你自己发一个版本就行了, 有空我会看一下你的代码。

**Reply by Jackhr-arch** [reply]
搞定了

理论上完全兼容v0.2版本的模板（因为用的一模一样的代码），同时允许添加一个`clashtui_template_version:1`作为模板标志（暂时未在生成的配置文件中添加类似的标志）

目前是使用profile区里所有的链接生成模板，后续可以慢慢改

看起来把no_proxy_provider的部分加上就可以正式发布v0.3.0了（当然，建议你测试一下各项功能）

**Reply by JohanChane** [reply]
看你的了
